apps/src/dance/Dance.js

```diff
@@ -141,10 +160,83 @@ Dance.prototype.loadAudio_ = function () {
   this.studioApp_.loadAudio(this.skin.failureSound, 'failure');
 };
 
+function p5KeyCodeFromArrow(idBtn) {
+  switch (idBtn) {
+    case ArrowIds.LEFT:
+      return window.p5.prototype.LEFT_ARROW;
+    case ArrowIds.RIGHT:
+      return window.p5.prototype.RIGHT_ARROW;
+    case ArrowIds.UP:
+      return window.p5.prototype.UP_ARROW;
+    case ArrowIds.DOWN:
+      return window.p5.prototype.DOWN_ARROW;
+  }
+}
+
+Dance.prototype.onArrowButtonDown = function (buttonId, e) {
+  // Store the most recent event type per-button
+  this.btnState[buttonId] = ButtonState.DOWN;
+  e.preventDefault();  // Stop normal events so we see mouseup later.
+
+  this.notifyKeyCodeDown(p5KeyCodeFromArrow(buttonId));
+};
+
+Dance.prototype.onArrowButtonUp = function (buttonId, e) {
+  // Store the most recent event type per-button
+  this.btnState[buttonId] = ButtonState.UP;
+
+  this.notifyKeyCodeUp(p5KeyCodeFromArrow(buttonId));
+};
+
+Dance.prototype.onMouseUp = function (e) {
+  // Reset all arrow buttons on "global mouse up" - this handles the case where
+  // the mouse moved off the arrow button and was released somewhere else
+
+  if (e.touches && e.touches.length > 0) {
+    return;
+  }
+
+  for (const buttonId in this.btnState) {
+    if (this.btnState[buttonId] === ButtonState.DOWN) {
+      this.onArrowButtonUp(buttonId, e);
+    }
+  }
+};
+
+Dance.prototype.notifyKeyCodeDown = function (keyCode) {
+  // Synthesize an event and send it to the internal p5 handler for keydown
+  if (this.p5) {
+    this.p5._onkeydown({ which: keyCode });
```

Comment:
Is there a way to dispatch this that doesn't require reaching in to private APIs?

Reply:
p5 was really intended to allow this. I'm not aware of another alternative approach. Synthesizing the key events with the expectation that p5 will see them and handle them identically to real key events seems riskier with various browsers.